### PR TITLE
[query] Lowered BlockMatrixBinaryReader and BlockMatrixBinaryWriter

### PR DESCRIPTION
**Pull request by johnc1231**
This allows the use of `from_numpy` and `to_numpy` on the local / service backends.

### CODE REVIEW

**Comment by johnc1231** [comment]
Problem with this one is that I have to use the right file api for service backend

**Comment by danking** [comment]
passes on local now: test_rectangles_to_numpy

I think the easiest thing to do is to just switch on spark vs non-spark. For spark its important we're not using HDFS for tofile, but for Local and Service, ideally, we just use a normal `hl.TemporaryFilename` and load it via `hl.current_backend().fs.open`. I'm not sure how finicky numpy.fromfile / tofile are about their file objects though

**Comment by johnc1231** [comment]
So yeah, there are two problems with this one. The second is that mysterious `test_rectangles_to_numpy` thing which shouldn't work and fails locally.

**Comment by danking** [comment]
I reproduced this by running all the tests together. There's some inter-test leakage.

**Comment by danking** [comment]
It does fail if I just run test_linalg.py, unfortunately.

**Comment by danking** [comment]
repro
```
make -C hail pytest PYTEST_ARGS="-k 'test_init_hail_context_twice or test_rectangles_to_numpy'" HAIL_QUERY_BACKEND=local
```

**Comment by danking** [comment]
Got it.

**Comment by johnc1231** [comment]
You can review Tim, mostly straightforward. The only thing I'm going to change is special casing on the backend to figure out where to write this numpy file

**Comment by tpoterba** [comment]
Looking now.

File: hail/src/main/scala/is/hail/expr/ir/functions/NDArrayFunctions.scala [hunk 1]
```diff
@@ -145,5 +145,11 @@ object  NDArrayFunctions extends RegistryFunctions {
         cb.ifx(info cne 0, cb._fatalWithError(errorID,s"hl.nd.solve: Could not solve, matrix was singular. dtrtrs error code ", info.toS))
         resPCode
     }
+
+    registerSCode2("export_doubles", TNDArray(TFloat64, Nat(2)), TString, TVoid, (t, ndSt, pathSt) => SVoid) { case (er, cb, rt, snd: SNDArrayValue, sstring: SStringValue, errorID) =>
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/functions/NDArrayFunctions.scala:149]: delete?

File: hail/src/main/scala/is/hail/types/encoded/ENumpyBinaryNDArray.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    cb.forLoop(cb.assign(i, 0L), i < n, cb.assign(i, i + 1L), {
+      readElemF(cb, region, currElementAddress, in)
+      cb.assign(currElementAddress, currElementAddress + pt.elementType.byteSize)
+    })
+
+    tFinisher(cb)
+  }
+
+  def _buildSkip(cb: EmitCodeBuilder, r: Value[Region], in: Value[InputBuffer]): Unit = {
+    ???
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/types/encoded/ENumpyBinaryNDArray.scala:64]: fine to go in like this.

**Review by tpoterba — CHANGES_REQUESTED** [review]
one piece of dead code to GC, otherwise looks great and elegant.

File: hail/src/main/scala/is/hail/types/encoded/ENumpyBinaryNDArray.scala (hunk 2 above)
Comment (johnc1231) [reply hail/src/main/scala/is/hail/types/encoded/ENumpyBinaryNDArray.scala:64]: It's never going to have a skipper. I should throw an error instead.

File: hail/src/main/scala/is/hail/expr/ir/functions/NDArrayFunctions.scala (hunk 1 above)
Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/functions/NDArrayFunctions.scala:149]: Yes will delete

**Comment by danking** [comment]
I pushed fixes for the filesystem issues.

**Review by danking — APPROVED** [review]
I addressed Tim's comment about deleting a function.

**Pull request merged**